File: src/experiment.py
import numpy as np
import click
import json

# convert numpy types to native python types
def convert_to_native(obj):
    if isinstance(obj, np.generic):
        return obj.item()
    elif isinstance(obj, dict):
        return {convert_to_native(k): convert_to_native(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [convert_to_native(i) for i in obj]
    else:
        return obj

# let the N people vote each other at random. with probability prob, a person does not vote
def compute(N, prob, rng):

    votes = np.zeros(N, dtype=int)
    abstained = 0
    for i in range(N):
        # if the random number is less than prob, the person does not vote
        if rng.random() < prob:
            abstained += 1
            continue
        else:
            # the person votes for another person (excluding i) at random
            eligible = set(range(N))
            eligible.remove(i)
            votes[rng.choice(list(eligible))] += 1

    return votes, abstained

# run the experiment for a single seed
def single_seed_experiment(rng, N, iterations, prob):

    # initialize the results arrays
    abstained_results = np.zeros(iterations, dtype=int)
    winner_results = np.zeros(iterations, dtype=int)

    print('Computing...')
    for i in range(iterations):
        # perform computation
        votes, abstained = compute(N, prob, rng)

        # save the results
        abstained_results[i] = abstained
        winner_results[i] = np.argmax(votes)

    # perform aggregation
    avg_num_abstained = np.mean(abstained_results)
    most_common_winner = np.argmax(np.bincount(winner_results))

    return {
        'avg_num_abstained': avg_num_abstained,
        'most_common_winner': most_common_winner
    }


# comment the click commands for testing
@click.command()
# the first three options are specified by the executor
@click.option('--exp_name', type=str, required=True, help="name of the experiment")
@click.option('--output_dir', type=str, required=True, help="output directory")
@click.option('--output_filename', type=str, required=True, help="output filename")
# the rest are experiment-specific
@click.option('--N', type=int, required=True, help="number of voters")
@click.option('--iterations', type=int, required=True, help="rounds of voting")
@click.option('--prob', type=float, default=0.1, help="probability of abstaining")
@click.option('--master_seed', type=int, default=42, help="the master seed for generating seeds")
@click.option('--num_seeds', type=int, default=0, help="number of derived seeds (0 = use master_seed directly)")
def experiment(exp_name, output_dir, output_filename, n, iterations, prob, master_seed, num_seeds):

    N = n  # click doesn't accept upper case arguments

    # make sure N>1
    if N < 2:
        raise ValueError('N must be greater than 1')

    # determine seeds to use
    if num_seeds <= 0:
        # single-seed mode: use master_seed directly
        seeds = [master_seed]
    else:
        # multi-seed mode: derive seeds from master_seed
        master_rng = np.random.default_rng(seed=master_seed)
        seeds = master_rng.integers(low=0, high=2**32-1, size=num_seeds)

    # collect seed results
    seed_results = []
    for seed in seeds:
        # initialize the random number generator
        rng = np.random.default_rng(seed=seed)
        # run the experiment
        result = single_seed_experiment(rng, N, iterations, prob)
        result['seed'] = int(seed)
        seed_results.append(result)

    # build flat output structure
    output = {
        'exp_name': exp_name,
        'N': N,
        'iterations': iterations,
        'prob': prob,
        'master_seed': master_seed,
        'num_seeds': num_seeds,
        'seed_results': seed_results
    }

    print('Converting to native Python types...')
    output = convert_to_native(output)

    print('Saving results...')
    with open(f'{output_dir}/{output_filename}', 'w') as f:
        json.dump(output, f, indent=2)

    print('Done!')

    return


if __name__ == '__main__':
    # comment for testing
    experiment()
    # uncomment for testing
    # experiment('configs/experiment.json', 0)

File: src/experiment.jl
using Random
using JSON
using Statistics
using ArgParse

# FIX: Pass rng to compute function to use seeded randomness
function compute(N, prob, rng)

    votes = zeros(Int, N)
    abstained = 0

    for i in 1:N
        if rand(rng) < prob
            abstained += 1
        else
            eligible = collect(setdiff(1:N, [i]))
            votes[rand(rng, eligible)] += 1
        end
    end

    return votes, abstained
end

# run the experiment for a single seed
function single_seed_experiment(rng, N, iterations, prob)

    # initialize results arrays
    abstained_results = zeros(Int, iterations)
    winner_results = zeros(Int, iterations)

    println("Computing...")
    for i in 1:iterations
        # perform computation (now passing rng)
        votes, abstained = compute(N, prob, rng)

        # save the results
        abstained_results[i] = abstained
        winner_results[i] = argmax(votes)
    end

    # perform aggregation
    avg_num_abstained = mean(abstained_results)
    num_of_wins = zeros(Int, N)
    for val in winner_results
        num_of_wins[val] += 1
    end
    most_common_winner = argmax(num_of_wins)

    return Dict(
        "avg_num_abstained" => avg_num_abstained,
        "most_common_winner" => most_common_winner
    )
end

function experiment(args)

    # parse the command line arguments
    s = ArgParseSettings()
    @add_arg_table! s begin
        "--exp_name"
            help="Name of the experiment"
            required=true
            arg_type=String
        "--output_dir"
            help="Output directory"
            required=true
            arg_type=String
        "--output_filename"
            help="Output filename"
            required=true
            arg_type=String
        "--N"
            help="Number of voters"
            required=true
            arg_type=Int
        "--iterations"
            help="Number of iterations"
            required=true
            arg_type=Int
        "--prob"
            help="Probability of abstaining"
            required=false
            default=0.1
            arg_type=Float64
        "--master_seed"
            help="The master seed for generating seeds"
            required=false
            default=42
            arg_type=Int
        "--num_seeds"
            help="Number of derived seeds (0 = use master_seed directly)"
            required=false
            default=0
            arg_type=Int
    end

    # parse the arguments
    parsed_args = parse_args(args, s)
    exp_name = parsed_args["exp_name"]
    output_dir = parsed_args["output_dir"]
    output_filename = parsed_args["output_filename"]
    N = parsed_args["N"]
    iterations = parsed_args["iterations"]
    prob = parsed_args["prob"]
    master_seed = parsed_args["master_seed"]
    num_seeds = parsed_args["num_seeds"]

    # make sure N>1
    if N < 2
        error("N must be greater than 1")
    end

    # determine seeds to use
    if num_seeds <= 0
        # single-seed mode: use master_seed directly
        seeds = [master_seed]
    else
        # multi-seed mode: derive seeds from master_seed
        master_rng = Random.Xoshiro(master_seed)
        seeds = rand(master_rng, UInt32, num_seeds)
    end

    # collect seed results
    seed_results = []
    for seed in seeds
        # initialize the random number generator
        rng = Xoshiro(seed)
        # run the experiment
        result = single_seed_experiment(rng, N, iterations, prob)
        result["seed"] = Int(seed)
        push!(seed_results, result)
    end

    # build flat output structure
    output = Dict(
        "exp_name" => exp_name,
        "N" => N,
        "iterations" => iterations,
        "prob" => prob,
        "master_seed" => master_seed,
        "num_seeds" => num_seeds,
        "seed_results" => seed_results
    )

    println("Saving results...")
    open(joinpath(output_dir, output_filename), "w") do f
        write(f, JSON.json(output, 2))
    end

    println("Done!")
end

# uncomment the following line for testing
experiment(ARGS)
